src/usr/local/www/firewall_rules.php

```diff
@@ -823,11 +823,25 @@ function($type) {
 							<?php endif; ?>
 						</td>
 						<td>
-							<?php if (isset($filterent['gateway'])): ?>
-								<span data-toggle="popover" data-trigger="hover focus" title="<?=gettext('Gateways details')?>" data-content="<?=gateway_info_popup($filterent['gateway'], $gateways_status)?>" data-html="true">
-							<?php else: ?>
-								<span>
-							<?php endif; ?>
+							<?php 
+								/* Cache gateway info for this page load.
+								 * See https://redmine.pfsense.org/issues/12174 */
+								if (isset($filterent['gateway'])) {
+									if (!is_array($gw_info)) {
+										$gw_info = array();
+									}
+									if (empty($gw_info[$filterent['gateway']])) {
+										$gw_info[$filterent['gateway']] = gateway_info_popup($filterent['gateway']);
+									}
+									if (!empty($gw_info[$filterent['gateway']])) {
+										echo sprintf('<span data-toggle="popover" data-trigger="hover focus" title="%s" data-content="%s" data-html="true">', gettext('Gateway details'), $gw_info[$filterent['gateway']]);
```

Comment:
I'd like to avoid doing a print/echo of HTML like this, it's better to do a similar style to what was removed where the logic is in PHP but the presentation is inline. It is much easier to maintain going forward and less prone to quirks like having to ensure things are quoted exactly right.

Reply:
@jim-p , I hope I did that right.  See the latest diff.

I presume you would keep your gateways_status cache as well, so I updated this function call to use it.